src/decompiler.cc

```diff
@@ -0,0 +1,426 @@
+        IndentValue(child, prefix.size(), prefix);
+        v.back().append(postfix.data(), postfix.size());
+      }
+    } else {
+      // Multiline with prefix on its own line.
+      IndentValue(child, indent_amount, {});
+      v.insert(v.begin(), std::string(prefix));
+      v.back().append(postfix.data(), postfix.size());
+    }
+  }
+
+  void BracketIfNeeded(Value &val) {
+    if (!val.needs_bracketing) return;
+    WrapChild(val, "(", ")");
+    val.needs_bracketing = false;
+  }
+
+  void WrapBinary(string_view infix, bool indent_right) {
+    auto right = std::move(stack.back());
+    stack.pop_back();
+    BracketIfNeeded(right);
+    auto left = std::move(stack.back());
+    stack.pop_back();
+    BracketIfNeeded(left);
+    auto width = infix.size() + left.width() + right.width();
+    if (width < target_exp_width && left.v.size() == 1 && right.v.size() == 1) {
+      stack.push_back({
+        { left.v[0] + std::string(infix) + right.v[0] },
+        true
+      });
+    } else {
+      Value bin { {}, true };
+      std::move(left.v.begin(), left.v.end(), std::back_inserter(bin.v));
+      bin.v.back().append(infix.data(), infix.size());
+      if (indent_right) IndentValue(right, indent_amount, {});
+      std::move(right.v.begin(), right.v.end(), std::back_inserter(bin.v));
+      stack.push_back(std::move(bin));
+    }
+  }
+
+  std::string TypeFromLoadStore(Opcode opcode, string_view name) {
+    auto op = std::string(OpcodeToToken(opcode));
+    auto load_pos = op.find(name.data(), 0, name.size());
+    if (load_pos != std::string::npos) {
+      op.erase(load_pos, name.size());
+      if (op.size() > load_pos) op.insert(load_pos, "_");
+    }
+    return op;
+  }
+
+  void DecompileExpr(const Expr& e) {
+    switch (e.type()) {
+      case ExprType::Const: {
+        auto& c = cast<ConstExpr>(&e)->const_;
+        switch (c.type) {
+          case Type::I32:
+            ss << static_cast<int32_t>(c.u32);
+            break;
+          case Type::I64:
+            ss << static_cast<int64_t>(c.u64) << "L";
+            break;
+          case Type::F32:
+            ss << *reinterpret_cast<const float *>(&c.f32_bits) << "f";
+            break;
+          case Type::F64:
+            ss << *reinterpret_cast<const double *>(&c.f64_bits) << "d";
+            break;
+          case Type::V128:
+            ss << "V128";  // FIXME
+            break;
+          default:
+            assert(false);
+        }
+        break;
+      }
+      case ExprType::LocalGet: {
+        auto& lg = *cast<LocalGetExpr>(&e);
+        ss << lg.var.name();
+        break;
+      }
+      case ExprType::GlobalGet: {
+        auto& gg = *cast<GlobalGetExpr>(&e);
+        ss << gg.var.name();
+        break;
+      }
+      case ExprType::LocalSet: {
+        auto& ls = *cast<LocalSetExpr>(&e);
+        WrapChild(stack.back(), ls.var.name() + " = ", "");
+        stack.back().needs_bracketing = true;
+        return;
```

Comment:
Oof, some of these breaking and some returning is weird.

Can we split this into two switches, one where we always early-return, another where we always break (plus the logic in `default`)? That makes the control flow edges less overlapping.

Or simpler, wrap the end of this function with `if (ss.str().size() > 0)` and always break. We return in the cases where we don't use `ss`, are there any times where we rely on keeping state in `ss` across calls to `DecompileExpr`?

Reply:
Agree, though this code is heavily in flux, so I want to avoid spending too much time perfecting it.

Actually, the simplest is to always make them return, and to stick the finishing code in its own function.